### ISSUE TITLE
all mutants escape, even if some should be caught

### ISSUE THREAD

**Issue by yivi** [infection/infection#322]
| Question    | Answer
| ------------| ---------------
| Infection version | 0.8.1 
| Test Framework version | 7.1.4
| PHP version | 7.2.4
| Platform    | MacOS
| Github Repo | - https://github.com/yivi/demu_mutant_fail

Created a very basic proof-of-concept project to play around with mutation testing.
There is a single class (`FooZero`), with a single method:

```php
public function zero(int $a): int

    {
        return $a * 0;
    }
```

And a single test covering it:
```php
 public function testGiveMeZero()
    {
        $a = new FooZero();
        $this->assertSame($a->zero(10), 0);
    }
```

The test passes normally: 

> vendor/bin/phpunit --testdox
> PHPUnit 7.1.4 by Sebastian Bergmann and contributors.
>
>Runtime:       PHP 7.2.4 with Xdebug 2.6.0
>Configuration: /Users/ivan.yivoff-ext/www_personal/demo_mutant_fail/phpunit.xml.dist
>
>FooMin
> ✔ Give me zero [7.49 ms]
>
> Time: 68 ms, Memory: 4.00MB

But when I run infection, it tells me that all mutants have escaped... 

```
vendor/bin/infection --test-framework-options="--bootstrap=bootstrap.php"
You are running Infection with xdebug enabled.
    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Running initial test suite...

PHPUnit version: 7.1.4

    7 [============================] < 1 sec

Generate mutants...

Processing source code files: 1/1
Creating mutated files and processes: 4/4
.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out

MMMM                                                 (4 / 4)

4 mutations were generated:
       0 mutants were killed
       0 mutants were not covered by tests
       4 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered
```
But I think that all four mutations should fail the test:

**PublicVisibility** should naturally fail.
**Multiplication** changes the zero multiplication for a zero division... which should logically fail.
**OneZeroInteger** would make the test fail, as a the method would return the number passed, and not zero. **DecrementInteger**  should fail similarly.

So my understanding would be that all mutants should have been caught and killed, but they have all escaped.

Linked github repo demoing issue.

**Comment by theofidry** [comment]
Hi, could you print the report as well please? There should be the details there

**Comment by yivi** [comment]
@BackEndTea Same result.

**Comment by yivi** [comment]
@theofidry You mean `infection-log.txt`? It's included in the repo.

I can copy it here as well:
```
Escaped mutants:
================


1) /Users/user/www_personal/demo_mutant_fail/src/FooZero.php:8    [M] DecrementInteger
exec /usr/local/Cellar/php/7.2.4_1/bin/php -c /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection8ka0oU  /Users/user/www_personal/demo_mutant_fail/vendor/phpunit/phpunit/phpunit --configuration /var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.15457aa30705ecdf657c596af6f32799.infection.xml --stop-on-failure --bootstrap=bootstrap.php
--- Original
+++ New
@@ @@
 {
     public function zero(int $a) : int
     {
-        return $a * 0;
+        return $a * -1;
     }
 }

PHPUnit 7.1.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.4
Configuration: /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.15457aa30705ecdf657c596af6f32799.infection.xml

.                                                                   1 / 1 (100%)

Time: 24 ms, Memory: 4.00MB

OK (1 test, 1 assertion)


2) /Users/user/www_personal/demo_mutant_fail/src/FooZero.php:8    [M] OneZeroInteger
exec /usr/local/Cellar/php/7.2.4_1/bin/php -c /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection8ka0oU  /Users/user/www_personal/demo_mutant_fail/vendor/phpunit/phpunit/phpunit --configuration /var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.b6e2381ea8005d6e651350e6f2913f7d.infection.xml --stop-on-failure --bootstrap=bootstrap.php
--- Original
+++ New
@@ @@
 {
     public function zero(int $a) : int
     {
-        return $a * 0;
+        return $a * 1;
     }
 }

PHPUnit 7.1.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.4
Configuration: /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.b6e2381ea8005d6e651350e6f2913f7d.infection.xml

.                                                                   1 / 1 (100%)

Time: 21 ms, Memory: 4.00MB

OK (1 test, 1 assertion)


3) /Users/user/www_personal/demo_mutant_fail/src/FooZero.php:8    [M] Multiplication
exec /usr/local/Cellar/php/7.2.4_1/bin/php -c /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection8ka0oU  /Users/user/www_personal/demo_mutant_fail/vendor/phpunit/phpunit/phpunit --configuration /var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.60c528ae4796f63ba12acf1e1544eeda.infection.xml --stop-on-failure --bootstrap=bootstrap.php
--- Original
+++ New
@@ @@
 {
     public function zero(int $a) : int
     {
-        return $a * 0;
+        return $a / 0;
     }
 }

PHPUnit 7.1.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.4
Configuration: /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.60c528ae4796f63ba12acf1e1544eeda.infection.xml

.                                                                   1 / 1 (100%)

Time: 20 ms, Memory: 4.00MB

OK (1 test, 1 assertion)


4) /Users/user/www_personal/demo_mutant_fail/src/FooZero.php:5    [M] PublicVisibility
exec /usr/local/Cellar/php/7.2.4_1/bin/php -c /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection8ka0oU  /Users/user/www_personal/demo_mutant_fail/vendor/phpunit/phpunit/phpunit --configuration /var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.7d8f63aab87b536cc0a99bc907c4e987.infection.xml --stop-on-failure --bootstrap=bootstrap.php
--- Original
+++ New
@@ @@
 
 class FooZero
 {
-    public function zero(int $a) : int
+    protected function zero(int $a) : int
     {
         return $a * 0;
     }
 }
PHPUnit 7.1.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.4
Configuration: /private/var/folders/vm/4njclq0551gfx1q36d_syk7sfhkp5k/T/infection/phpunitConfiguration.7d8f63aab87b536cc0a99bc907c4e987.infection.xml

.                                                                   1 / 1 (100%)

Time: 21 ms, Memory: 4.00MB

OK (1 test, 1 assertion)

Timeout mutants:
================

Killed mutants:
===============

Errors mutants:
===============

Not covered mutants:
====================
```

**Comment by yivi** [comment]
But the bootstrap file is read thanks to the `test-framework-options` parameter. Otherwise tests would fail because they would be able to instantiate try class. 

I’ve actually tried that, and all mutants get (wrongly) killed because tests fail when not loading my bootstrap file.

**Comment by Fenikkusu** [comment]
One thing to consider is that by default, division by zero is a warning, not fatal error. You likely want to update your configuration such that any error/output is considered a failed test (through php). Next, I think there needs to be some clarification, as I still think you will run into some escaped mutants.

PublicVisibility - This is still going to happen still, until you change it to private instead of public or protected.
Multiplication - This will change it to a division, which unless your testing is configured properly, will still be an escaped mutant.
DecrementInteger - This would should get killed.
OneZeroInteger - This two should get killed.

**Comment by yivi** [comment]
I thought that under PHP7 division by zero became an exception, but I see that is not thrown reliably.

Still, in this mutation:
```php
 {
     public function zero(int $a) : int
     {
-        return $a * 0;
+        return $a / 0;
     }
 }
```
Should result on a killed mutant. When I try the test with that mutation in place, I get:
```
There was 1 error:

1) FooMinTest::testGiveMeZero
Division by zero
```
And even if for some reason id didn't catch the division by zero, it should fail as well. Since the result of `a / 0` wouldn't be an int, and wouldn't match the method signature. The test would fail.

But still infection reports an escaped mutant.

**PublicVisibility**, this mutation does the following:

```
-    public function zero(int $a) : int
+    protected function zero(int $a) : int
     {
         return $a * 0;
     }
```

Considering that the test does:
```
 $this->assertSame(0, $a->zero(100));
```
It should never pass with that mutation in place. If zero is protected, we no longer could do `$a->zero()`, and the mutant should die.

So I still think the four mutations should be killed. Just tried it again in a different machine, this one with PHP7.1, and I get the same result.

**Comment by borNfreee** [comment]
Thanks for creating this issue @yivi! Especially thanks for the repository with reproducible bug.

It's now fixed in https://github.com/infection/infection/pull/323, I will let you know when it's merged.

PS: please, consider using PSR-4 autoloading with composer rather than custom bootstrap/autoloader.

**Comment by borNfreee** [comment]
Fixed in #323
